opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/util/AttributeWriter.java

```diff
@@ -321,24 +318,73 @@ private static void validateArrayType(
+                    new NodeId(nodeMap.getNamespaceTable().getIndex(NamespaceTable.OPC_UA_NAMESPACE),
+                        valueDataTypeId)).get();
+                if (dataTypeId.equals(builtInTypeId.getNodeId()) ||
+                    subtypeOf(nodeMap, builtInTypeId.getNodeId(), dataTypeId)) {
+                    return valueClass;
+                } else {
+                    throw new UaException(StatusCodes.Bad_TypeMismatch);
+                }
+            } else {
+                throw new UaException(StatusCodes.Bad_TypeMismatch);
+            }
+        }
+    }
+
+    private static boolean subtypeOf(ServerNodeMap nodeMap, NodeId dataTypeId, NodeId parentType) {
+        ServerNode dataTypeNode = nodeMap.get(dataTypeId);
+        if (dataTypeNode != null) {
+            Optional<NodeId> superTypeId = getParentReference(nodeMap, dataTypeId, false);
+            return superTypeId.map(id -> id.equals(parentType) || subtypeOf(nodeMap, id, parentType)).orElse(false);
+        } else {
+            return false;
+        }
+    }
+
+    private static NodeId getBuiltInParent(ServerNodeMap nodeMap, NodeId dataTypeId) {
         ServerNode dataTypeNode = nodeMap.get(dataTypeId);
+        if (dataTypeNode != null) {
+            Optional<NodeId> superTypeId = getParentReference(nodeMap, dataTypeId, true);
+            return getParentReference(nodeMap, dataTypeId, true).isPresent() ? superTypeId.get() : null;
+        }
+        return null;
+    }
 
+    private static Optional<NodeId> getParentReference(ServerNodeMap nodeMap, NodeId dataTypeId, boolean builtinCheck) {
+        ServerNode dataTypeNode = nodeMap.get(dataTypeId);
         if (dataTypeNode != null) {
             Optional<NodeId> superTypeId = dataTypeNode.getReferences().stream()
-                .filter(r ->
-                    r.getReferenceTypeId().equals(Identifiers.HasSubtype) &&
-                        r.isInverse() &&
-                        r.getTargetNodeClass() == NodeClass.DataType)
+                .filter(r -> referenceFilter(r, builtinCheck))
                 .flatMap(r -> opt2stream(r.getTargetNodeId().local()))
                 .findFirst();
-
-            return superTypeId
-                .map(id -> id.equals(Identifiers.Structure) || isStructureSubtype(nodeMap, id))
-                .orElse(false);
-        } else {
-            return false;
+            return superTypeId;
         }
+        return null;
     }
 
-}
+    private static boolean referenceFilter(Reference r, boolean builtinCheck) {
+        boolean ret = r.getReferenceTypeId().equals(Identifiers.HasSubtype) &&
+            r.getTargetNodeClass() == NodeClass.DataType &&
+            r.isInverse();
+        return builtinCheck ? ret && TypeUtil.isBuiltin(r.getTargetNodeId()) : ret;
+    }
+
+    private static boolean hasBuiltInParentType(ServerNodeMap serverNodeMap, NodeId dataTypeId) {
+        return getBuiltInParent(serverNodeMap, dataTypeId) != null;
+    }
+}
```

Comment:
Newline at end of files. There's a setting for this; not sure if I can get checkstyle to enforce it or not. Sounds like something it should be able to do...